code/game/objects/items/devices/moodlight.dm

```diff
@@ -0,0 +1,111 @@
+	name = "RGB Moodlight"
+	desc = "A customizable RGB LED Moodlight that smoothly fades between colors. Perfect for SpaceTok!"
+	icon = 'icons/obj/device.dmi'
+	icon_state = "moodlight_off"
+	item_state = "moodlight_off"
+	throwforce = 0
+	w_class = WEIGHT_CLASS_SMALL
+	throw_speed = 3
+	throw_range = 4
+	var/activated = FALSE
+	var/list/colors = list()
+	var/list/choices = list("Toggle Power", "Add a color", "Remove a color", "Set Range", "Set Brightness")
+	var/first_color	//This is the current color we are on, not the first color in the list.
+	var/second_color //This is the color we are currently shifting to, from first_color
+	var/amount_shifted //This should always be a decimal number from 0 to 1 as this is used for the percentage in BlendRGB.
+	var/i = 1 //Iteration. Example: Shifting from color 1 to color 2, then shifting from color 2 to color 3. That would be 2 iterations.
+	var/range = 4
+	var/brightness = 2
+
+/obj/item/moodlight/attack_self(mob/user)
+	var/choice = input(user, "What do you want to do?") as null|anything in sortTim(choices, /proc/cmp_text_asc)
+	switch(choice)
+		if("Toggle Power")
+			if(length(colors) == 0)
+				to_chat(user, "<span class = 'warning'>Select some colors first!</span>")
+				return
+			toggle_shift(!activated)
+			visible_message("<span class = 'notice'>[user] turns [activated ? "on" : "off"] the [src]!</span>")
+		if("Add a color")
+			if(activated)
+				to_chat(user, "<span class = 'warning'>Turn it off first!</span>")
+				return
+
+			var/new_color = input(user, "Please select a color") as color
+			colors += new_color
+		if("Remove a color")
+			if(activated)
+				to_chat(user, "<span class = 'warning'>Turn it off first!</span>")
+				return
+
+			var/remove_this_color = input(user, "Please select a color") as null|anything in colors
+			if(!remove_this_color) //they cancelled
+				return
+			colors -= remove_this_color
+		if("Set Range")
+			var/new_range = input(user, "Please input a range from 1-6. Default is 4.") as num
+			if(new_range > 6 || new_range < 1)
+				to_chat(user, "<span class = 'warning'>Choose a level between 1-6!</span>")
+				return
+			range = new_range
+		if("Set Brightness")
+			var/new_brightness = input(user, "Please input a brightness from 0-2. Default is 2. Decimals allowed.") as num
+			if(new_brightness > 2 || new_brightness <= 0)
+				to_chat(user, "<span class = 'warning'>Choose a level between 0-2!</span>")
+				return
+			brightness = new_brightness
+		
+
+/obj/item/moodlight/extinguish_light() //for shadowlings
+	toggle_shift(FALSE)
+
+/obj/item/moodlight/proc/toggle_shift(active)
+	first_color = colors[1]
+	second_color = colors[2]
+	icon_state = "moodlight_[active ? "on" : "off"]"
+	item_state = "moodlight_[active ? "on" : "off"]"
+	activated = active
+	if(length(colors) == 1) //If there is only 1 color we dont need to process anything
+		set_light(range, brightness, colors[1])
+		color = colors[1]
+		return
+	if(active)
+		START_PROCESSING(SSfastprocess, src)
```

Comment:
What does it look like if you just use the normal processing and multiply your `amount_shifted` value by 20? Fast processing takes up a lot of processing speed so it would be good to minimise its usage if possible

Reply:
I have changed it to SSProcessing, it's not as smooth now but it's smooth enough